L1Trigger/L1CaloTrigger/plugins/Phase2L1CaloPFClusterEmulator.cc

```diff
@@ -0,0 +1,190 @@
+// Class:      Phase2L1CaloPFClusterEmulator
+//
+/**\class Phase2L1CaloPFClusterEmulator Phase2L1CaloPFClusterEmulator.cc L1Trigger/L1CaloTrigger/plugins/Phase2L1CaloPFClusterEmulator.cc
+
+ Description: Creates 3x3 PF clusters from GCTintTowers to be sent to correlator. Follows firmware logic, creates 8 clusters per (2+17+2)x(2+4+2).
+
+ Implementation: To be run together with Phase2L1CaloEGammaEmulator.
+
+*/
+//
+// Original Author:  Pallabi Das
+//         Created:  Wed, 21 Sep 2022 14:54:20 GMT
+//
+//
+
+// system include files
+#include <memory>
+
+// user include files
+#include "FWCore/Framework/interface/Frameworkfwd.h"
+#include "FWCore/Framework/interface/stream/EDProducer.h"
+
+#include "FWCore/Framework/interface/Event.h"
+#include "FWCore/Framework/interface/MakerMacros.h"
+
+#include "FWCore/ParameterSet/interface/ParameterSet.h"
+#include "FWCore/Utilities/interface/StreamID.h"
+
+#include "DataFormats/L1TCalorimeterPhase2/interface/CaloCrystalCluster.h"
+#include "DataFormats/L1TCalorimeterPhase2/interface/CaloTower.h"
+#include "DataFormats/L1TCalorimeterPhase2/interface/CaloPFCluster.h"
+#include "DataFormats/L1Trigger/interface/EGamma.h"
+
+#include <ap_int.h>
+#include <fstream>
+#include <iomanip>
+#include <iostream>
+#include <stdio.h>
+#include "L1Trigger/L1CaloTrigger/interface/Phase2L1CaloPFClusterEmulator.h"
+
+//
+// class declaration
+//
+
+class Phase2L1CaloPFClusterEmulator : public edm::stream::EDProducer<> {
+public:
+  explicit Phase2L1CaloPFClusterEmulator(const edm::ParameterSet&);
+  ~Phase2L1CaloPFClusterEmulator();
+
+  static void fillDescriptions(edm::ConfigurationDescriptions& descriptions);
+
+private:
+  void produce(edm::Event&, const edm::EventSetup&) override;
+
+  // ----------member data ---------------------------
+  edm::EDGetTokenT<l1tp2::CaloTowerCollection> caloTowerToken_;
+};
+
+//
+// constants, enums and typedefs
+//
+
+//
+// static data member definitions
+//
+
+//
+// constructors and destructor
+//
+Phase2L1CaloPFClusterEmulator::Phase2L1CaloPFClusterEmulator(const edm::ParameterSet& iConfig)
+    : caloTowerToken_(consumes<l1tp2::CaloTowerCollection>(iConfig.getParameter<edm::InputTag>("gctFullTowers"))) {
+  produces<l1tp2::CaloPFClusterCollection>("GCTPFCluster");
+}
+
+Phase2L1CaloPFClusterEmulator::~Phase2L1CaloPFClusterEmulator() {}
+
+//
+// member functions
+//
+
+// ------------ method called to produce the data  ------------
+void Phase2L1CaloPFClusterEmulator::produce(edm::Event& iEvent, const edm::EventSetup& iSetup) {
+  using namespace edm;
+  std::unique_ptr<l1tp2::CaloPFClusterCollection> pfclusterCands(make_unique<l1tp2::CaloPFClusterCollection>());
+
+  edm::Handle<std::vector<l1tp2::CaloTower>> caloTowerCollection;
+  if (!iEvent.getByToken(caloTowerToken_, caloTowerCollection))
+    edm::LogError("Phase2L1CaloPFClusterEmulator") << "Failed to get towers from caloTowerCollection!";
+
+  iEvent.getByToken(caloTowerToken_, caloTowerCollection);
```

Comment:
You check if you can get this, and log an error... but just proceed onward to try and get this anyways? Is this module tolerant to this error? Should this throw a `cms::exception` instead if these are not available?